PSYKHANOOL/PSYKHANOOL.h

```diff
@@ -0,0 +1,664 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// Current-friendly data manipulation types:
+//
+// * bool Exists(x) : non-throwing, always `true` for non-optionals.
+// * T Value(x)     : can throw `NoValue` if and only if `Exists(x)` is `false`.
+//
+// Current-friendly container types.
+//
+// * Vector<T> <=> std::vector<T>
+//   Empty(), Size(), operator[](i), PushBack(x), PopBack() [, iteration].
+//
+// * OrderedDictionary<T> <=> std::map<T_KEY, T>
+//   Empty(), Size(), operator[](key), Erase(key) [, iteration, {lower/upper}_bound].
+//   `T_KEY` is either the type of `T.key` or of `T.get_key()`.
+//
+// * LightweightMatrix<T> <=> { T_ROW, T_COL } -> T, three `std::map<>`-s.
+//   Empty(), Size(), Rows()/Cols(), Add(cell), Delete(row, col) [, iteration, {lower/upper}_bound].
+//   `T_ROW` and `T_COL` are either the type of `T.row` / `T.col`, or of `T.get_row()` / `T.get_col()`.
+//
+// ...
+//
+// All Current-friendly types support persistence.
+//
+// Only allow default constructors for containers.
+
+#ifndef PSYKHANOOL_H
+#define PSYKHANOOL_H
+
+#include <fstream>
+#include <vector>
+#include <map>
+#include <utility>
+
+#include "sfinae.h"
+
+#include "../Bricks/time/chrono.h"
+#include "../Bricks/strings/strings.h"
+#include "../Bricks/cerealize/cerealize.h"
+
+namespace PSYKHANOOL {
+
+struct NoValueException : std::exception {};
+typedef const NoValueException& NoValue;
+
+struct CannotPopBackFromEmptyVectorException : std::exception {};
+typedef const CannotPopBackFromEmptyVectorException& CannotPopBackFromEmptyVector;
+
+template <typename T>
+class ImmutableOptional final {
+ public:
+  ImmutableOptional() = delete;
+  ImmutableOptional(const T* object) : optional_object_(object) {}
+  ImmutableOptional(std::unique_ptr<T>&& rhs)
+      : owned_optional_object_(std::move(rhs)), optional_object_(owned_optional_object_.get()) {}
+  bool Exists() const { return optional_object_ != nullptr; }
+  const T& Value() const {
+    if (optional_object_ != nullptr) {
+      return *optional_object_;
+    } else {
+      throw NoValueException();  // TBD: Derived detailed exception type.
+    }
+  }
+
+ private:
+  std::unique_ptr<T> owned_optional_object_;
+  const T* optional_object_;
+};
+
+template <typename T>
+class MutableOptional final {
+ public:
+  MutableOptional() = delete;
+  MutableOptional(T* object) : optional_object_(object) {}
```

Comment:
What's the difference between `ImmutableOtional` and `MutableOptional` with respect to passing `unique_ptr` to them?